### ISSUE TITLE
Allow installing additional dependencies in Docker deployment

### ISSUE THREAD

**Issue by jezdez** [getredash/redash#2810]
### Issue Summary

To be able to use the Python extension API (and soon the frontend extension API) it would be useful to be able to easily install extra dependencies when using the Redash docker image. I wonder if we could add an optional environment variables, e.g. `REDASH_EXTRA_PYTHON_REQUIREMENTS` and the analog way for Node dependencies, to be picked up when inheriting from the main Redash Docker image.

It's simple enough to do that right now by simply inheriting from the Redash image and running additional commands, e.g.:

```dockerfile
FROM redash/redash:latest

COPY requirements_extra.txt /tmp/
RUN pip install -r /tmp/requirements_extra.txt
```

If we want to encourage users to use the extension APIs in their projects, a guideline how to integrate it in the Docker based deployment workflow would alleviate the risk of the application bundle build process from breaking.

The above example is simple enough, but when considering that the [current Dockerfile](https://github.com/getredash/redash/blob/93b34bcbce2b3df328e1c1fac4261fa83e8717b3/Dockerfile#L6-L9) installs the Node dependencies only to run build the application bundle and then deletes `node_modules/` again, it's a lot more brittle that I think is worthwhile going forward.

For the sake of consistency between the Python and Node dependencies it would make sense to have a common API to add additional dependencies using the parent Redash Docker image.

Maybe an ideal way would be to use the [`ON BUILD`](https://docs.docker.com/engine/reference/builder/#onbuild) command to do the installation and build process if certain environment variables are defined? OTOH the `*-onbuild` images of the [various language images have recently been deprecated](https://github.com/docker-library/official-images/issues/2076), so maybe this is betting on the wrong horse.

Also mentioned in that ticket, Sentry uses `ON BUILD` in their ["onpremise" Docker image](https://github.com/getsentry/onpremise/blob/master/Dockerfile) though, which seems to be an accepted use case. The base onbuild image contains a few [`ON BUILD` commands to install additional dependencies if available](https://github.com/getsentry/docker-sentry/blob/f648178adf33bec68be47550a82850e18deb0cd8/9.0/onbuild/Dockerfile#L9-L13):

```dockerfile
FROM sentry:9.0

# ...

# Hook for installing additional plugins
ONBUILD RUN if [ -s requirements.txt ]; then pip install -r requirements.txt; fi

# Hook for installing a local app as an addon
ONBUILD RUN if [ -s setup.py ]; then pip install -e .; fi
```

I'm not sure what the best way would be to do the same for NPM though since we'll need to install both Redash's own **and** custom dependencies at the same time to successfully build the application bundle.

### Technical details:

* Redash Version: 5.0beta
* Browser/OS: Firefox/macOS
* How did you install Redash: Docker

**Comment by arikfr** [comment]
If I understand it correctly, what ONBUILD allows is running additional commands with a new context (files) _after_ the original image was built already. It only saves the end user the need to write some commands and adds standardization, but won't change how things are built.

TL;DR it's only useful for installing additional Python packages, but not for NPM/bundling the frontend assets.

I don't mind adding this extensions to make things easier for all involved.

As a side note, on the topic of improving our Dockerfile, I was thinking of switching to a [multi stage Docker build](https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds) to optimize the final size and speed up things (by not having to install Node utilities for example but using a Node Docker image).
